File: examples/python/sdk-reference/error-handling/with-retry-helper.py
from aws_durable_execution_sdk_python.config import Duration
from aws_durable_execution_sdk_python.context import DurableContext
from aws_durable_execution_sdk_python.execution import durable_execution
from aws_durable_execution_sdk_python.retries import (
    RetryStrategyConfig,
    WithRetryConfig,
    create_retry_strategy,
    with_retry,
)


def charge_flow(ctx: DurableContext, attempt: int) -> str:
    # invoke does not accept a retry strategy, so with_retry handles backoff.
    return ctx.invoke(
        function_name="process-payment",
        payload={"order_id": "abc"},
        name=f"charge-{attempt}",
    )


retry_config = WithRetryConfig(
    retry_strategy=create_retry_strategy(
        RetryStrategyConfig(
            max_attempts=3,
            initial_delay=Duration.from_seconds(2),
            backoff_rate=2.0,
        )
    ),
)


@durable_execution
def lambda_handler(event: dict, context: DurableContext) -> str:
    return with_retry(
        context,
        func=charge_flow,
        config=retry_config,
        name="charge-payment",
    )

File: examples/csharp/sdk-reference/error-handling/with-retry-helper.cs
using Amazon.Lambda.Core;
using Amazon.Lambda.DurableExecution;

public class WithRetryHelperExample
{
    public Task<DurableExecutionInvocationOutput> Handler(
        DurableExecutionInvocationInput input, ILambdaContext context)
        => DurableFunction.WrapAsync<OrderEvent, string>(Workflow, input, context);

    private async Task<string> Workflow(OrderEvent input, IDurableContext ctx)
    {
        // .NET has no separate withRetry helper. InvokeConfig does not accept a retry
        // strategy, so wrap the invoke in a step and set the retry strategy on StepConfig.
        // The step retries the invocation with backoff between failed attempts.
        var config = new StepConfig
        {
            RetryStrategy = RetryStrategy.Exponential(
                maxAttempts: 3,
                initialDelay: TimeSpan.FromSeconds(2),
                maxDelay: TimeSpan.FromMinutes(1),
                backoffRate: 2.0,
                jitter: JitterStrategy.Full),
        };

        return await ctx.StepAsync(
            async (_, ct) => await ctx.InvokeAsync<PaymentRequest, string>(
                "process-payment",
                new PaymentRequest(input.OrderId),
                name: "charge",
                cancellationToken: ct),
            name: "charge-payment",
            config: config);
    }
}

public record OrderEvent(string OrderId);
public record PaymentRequest(string OrderId);
